include/roboteam_utils/containers/state_machine.hpp

```diff
@@ -0,0 +1,213 @@
+             * Now say you'd want to add a space:
+             * ((std::cout << args << " "), ...);
+             *
+             * Add up a pack of ints together?
+             *
+             * int result = (args + ...);
+             */
+            (_data.emplace_back(std::make_unique<Tys>(std::forward<Tys>(args))), ...);
+        }
+
+        /**
+         * Gets a mutable iterator to the first element
+         * @return std::vector<Base>::iterator
+         */
+        iterator begin() noexcept {
+            return _data.begin();
+        }
+
+        /**
+         * Gets a mutable iterator to 1 past the last element
+         * @return std::vector<Base>::iterator
+         */
+        iterator end() noexcept {
+            return _data.end();
+        }
+
+        /**
+         * Gets an immutable iterator to the first element
+         * @return std::vector<Base>::const_iterator
+         */
+        const_iterator begin() const noexcept {
+            return _data.cbegin();
+        }
+
+        /**
+         * Gets an immutable iterator to 1 past the last element
+         * @return std::vector<Base>::const_iterator
+         */
+        const_iterator end() const noexcept {
+            return _data.cend();
+        }
+
+        /**
+         * Skips n elements, sets current to clamp(current + n, _data.size())
+         * @param n Amount of elements to skip
+         */
+        void skip_n(size_t n) noexcept {
+            current += n;
+            current = std::clamp<size_t>(current, 0, _data.size());
+        }
+
+        /**
+         * Checks whether al elements have been finalized
+         * @return current_num() == total_count();
+         */
+        [[nodiscard]] bool finished() const noexcept {
+            return current_num() == total_count();
+        }
+
+        /**
+         * Gets current index into state machines
+         * @return this->current
+         */
+        [[nodiscard]] size_t current_num() const noexcept {
+            return current;
+        }
+
+        /**
+         * Gets total amount of elements in the state machine
+         * @return vector.size();
+         */
+        [[nodiscard]] size_type total_count() const noexcept {
+            return _data.size();
+        }
+
+        /**
+         * Gets the amount of elements in the vector
+         * @return vector.size()
+         */
+        [[nodiscard]] size_type size() const noexcept {
+            return _data.size();
+        }
+
+        /**
+         * Updates the current element in the vector
+         * @param data Data to pass to .update(data);
+         * @return .update(data)
+         */
+        [[nodiscard]] Enum update(SI const& data) noexcept {
+            return begin()[current_num()]->update(data);
+        }
+
+        /**
+         * Calls terminate() on the current Base
+         * begin(0[current_num()]->terminate();
+         */
+        void terminate() noexcept {
+            begin()[current_num()]->terminate();
```

Comment:
Why did you choose to implement it this way instead of getcurrent()->terminate?

Reply:
because I made this before i made get_current()